### ISSUE TITLE
Puma server not restarting after a deployment

### ISSUE THREAD

**Issue by darkcode85** [ajgon/opsworks_ruby#254]
### Stack

```
OS: Ubuntu 18.04 LTS
cookbook 'opsworks_ruby', '~> 1.20.2'
cookbook 'packages', '~> 1.0.0'
```

### Expected Behavior
 
When deployment is marked as completed in the OpsWorks UI the code changes should be reflected automatically.

_We are using the latest version of opsworks_ruby cookbook (`1.20.2`)_

### Actual Behavior

After a deployment has been completed and OpsWorks UI shows the deployment as completed the changes are not reflected and the application still shows the previous code before the deployment, we have to ssh login into the server and run `monit restart puma_app_name` which then force the new changes to be reflected.

However in the deployments logs we can see that `monit restart` is executed for puma and sidekiq.

#### Deployment logs:

```
...
[2021-02-24T01:20:05+00:00] INFO: deploy[app_name] deployed to /srv/www/app_name
...
[2021-02-24T01:20:05+00:00] INFO: Processing execute[monit restart puma_app_name] action run (opsworks_ruby::deploy line 55)
[2021-02-24T01:20:05+00:00] INFO: execute[monit restart puma_app_name] ran successfully
[2021-02-24T01:20:05+00:00] INFO: Processing execute[monit restart sidekiq_app_name-1] action run (opsworks_ruby::deploy line 38)
[2021-02-24T01:20:05+00:00] INFO: execute[monit restart sidekiq_app_name-1] ran successfully
[2021-02-24T01:20:05+00:00] INFO: Processing execute[monit restart sidekiq_app_name-2] action run (opsworks_ruby::deploy line 38)
[2021-02-24T01:20:05+00:00] INFO: execute[monit restart sidekiq_app_name-2] ran successfully
[2021-02-24T01:20:05+00:00] INFO: deploy[app_name] sending reload action to service[nginx] (delayed)
[2021-02-24T01:20:05+00:00] INFO: Processing service[nginx] action reload (opsworks_ruby::deploy line 44)
[2021-02-24T01:20:05+00:00] INFO: service[nginx] reloaded
[2021-02-24T01:20:06+00:00] INFO: HTTP Request Returned 404 Not Found: Object not found: chefzero://localhost:8889/nodes/rails-app-server2
[2021-02-24T01:20:06+00:00] INFO: Chef Run complete in 170.842638575 seconds
[2021-02-24T01:20:06+00:00] INFO: Running report handlers
[2021-02-24T01:20:06+00:00] INFO: Report handlers complete
```

### Custom JSON
```json
{
    "ruby-version": "2.5",
    "use-nodejs": true,
    "deploy": {
        "app_name": {
            "global": {
                "environment": "production"
            },
            "webserver": {
                "adapter": "nginx",
                "enable_upgrade_method": true
            },
            "appserver": {
                "adapter": "puma",
                "application_yml": true
            },
            "worker": {
                "adapter": "sidekiq"
            }
        }
    }
}
```

**Comment by darkcode85** [comment]
After a few days and deploying the app multiples times I have notice this error is inconsistent, sometimes the deployments changes are reflected after a few seconds, other times after a few minutes and sometimes the changes are never reflected until puma is restarted manually.

**Comment by inopinatus** [comment]
@darkcode85 that sounds very similar to the misbehaviours I've observed for unicorn in #256.

I'm coming to the view that

a) monit should never be stopping and restarting the application server in normal use; the post-deploy command should be a refork from a master process, and

b) monit expects the "start program" stanza to return quickly. If the "start program" command runs the application server in the foreground, this eventually looks like a failure by timeout to Monit, which therefore becomes confused and is the root cause of these failure modes.

See also https://github.com/puma/puma/issues/1983.

**Comment by olbrich** [comment]
I have been running into issues where during a deploy requests to the server can be completely dropped as monit stops and then starts puma. It appears that when we moved to a foreground process, we lost the ability to do proper restarts and rolling restarts (which are usually done by issuing a USR2 signal instead of killing and restarting puma).  

I also just noticed that we have lost the appserver service scripts, which were occasionally useful. I'm not entirely sure, but I suspect that the different app servers each have different ways of restarting, so there may be a need to have appserver specific monit scripts or to bring back the appserver specific service scripts and have monit invoke the appropriate one instead.
